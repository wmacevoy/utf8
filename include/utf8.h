#pragma once
// input
//
//   utf8 - encoded string of length n (may contain 0's)
//   n    - length of utf8 string
//
// return value
//
//   1-4 length of utf8-encoded codepoint as first character of utf8
//   -1  invalid encoding (or n <= 0)

#include <stddef.h>

#ifndef UTF8_C
extern unsigned UTF8_MINVAL[4];
extern unsigned UTF8_MAXVAL[4];
#endif

int utf8declen(const unsigned char *utf8, int n);
int utf8decval(const unsigned char *utf8, int len);
int utf8enclen(unsigned val);
void utf8encval(unsigned char *utf8, unsigned val, int enclen);

int utf8codecount(const unsigned char *utf8, int utf8len);
int utf8codewrite(const unsigned char *utf8, int utf8len, wchar_t *dest, int destlen);

