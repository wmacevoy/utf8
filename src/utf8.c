#define UTF8_C
#include "utf8.h"

unsigned UTF8_MINVAL[4] = { 0x00, 0x80,   0x800,  0x10000 };
unsigned UTF8_MAXVAL[4] = { 0x7F, 0x7FF, 0xFFFF, 0x1FFFFF };

int utf8declen(const unsigned char *utf8, int n) {
  if (n <= 0) return -1;
  unsigned char c = utf8[0];
  if ((c & 0x80) == 0) return 1;
  if ((c & 0x40) != 0) {
    for (int len=2; len <= 4; ++len) {
      if (n < len || ((utf8[len-1] & 0xC0) != 0x80)) return -1;

      if ((c & (0x80 >> len)) == 0) return len;
    }
  }
  return -1;
}

int utf8decval(const unsigned char *utf8, int len) {
  if (len == 1) {
    return utf8[0];
  } else if (len > 0) {
    unsigned val = utf8[0] & ~(0xFF << (7-len));
    for (int i=1; i<len; ++i) {
      val = (val  << 6) | (utf8[i] & 0x3F);
    }
    return val;
  }
  return -1;
}

int utf8enclen(unsigned val) {
  if (val < 0x80) {
    return 1;
  } else if (val < 0x0800) {
    return 2;
  } else if (val < 0x10000) {
    return 3;
  }
  return 4;
}

static unsigned char utf8encmask[] = {0x00,0xC0,0xE0,0xF0};

void utf8encval(unsigned char *utf8, unsigned val, int enclen) {
  for (int i=enclen-1; i>0; --i) {
    utf8[i]=(0x80)|(val&0x3F);
    val >>= 6;
  }
  utf8[0]=utf8encmask[enclen-1]|val;
}

int utf8decode(const unsigned char *utf8, int utf8len, wchar_t *wc, int wccap) {
  int ans = 0;
  while (utf8len > 0) {
    int declen=utf8declen(utf8,utf8len);
    if (declen > 0) {
      if (wccap > 0) {
	wc[0]=utf8decval(utf8,declen);
	--wccap;
	++wc;
      }
      utf8 += declen;
      utf8len -= declen;
      ans += 1;
    } else {
      if (wccap > 0) {
	wc[0]=utf8[0];
	--wccap;
	++wc;
      }
      utf8 += 1;
      utf8len -= 1;
      ans += 1;
    }
  }
  return ans;
}

int utf8encode(const wchar_t *wc, int wclen, unsigned char *utf8, int utf8cap) {
  int ans = 0;
  while (wclen > 0) {
    int enclen=utf8enclen(wc[0]);
    ans += enclen;
    if (enclen <= utf8cap) {
	utf8encval(utf8, wc[0], enclen);
	utf8cap -= enclen;
	utf8 += enclen;
    } else {
      utf8cap = 0; // stop further writes
    }
    wc += 1;
    wclen -= 1;
  }
  return ans;
}

