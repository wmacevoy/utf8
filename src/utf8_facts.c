
#include <stdlib.h>
#include <stdio.h>

#include "facts.h"
#include "utf8.h"

unsigned char b(const char *str) {
  unsigned char x = 0;
  while (*str != 0) {
    if (*str == '0' || *str == '1') {
      x = (x << 1) | (*str == '1');
    }
    ++str;
  }
  return x;
}

void utf8teststr(unsigned char *dst,int len,int ab) {
  switch(len) {
  case 1:
    if (ab == 0) { snprintf(dst,len+1,"%c",b("0000-0000")); }
    if (ab == 1) { snprintf(dst,len+1,"%c",b("0111-1111")); }
    break;
  case 2:
    if (ab == 0) { snprintf(dst,len+1,"%c%c",b("1100-0010"),b("1000-0000")); }
    if (ab == 1) { snprintf(dst,len+1,"%c%c",b("1101-1111"),b("1011-1111")); }
    break;  
  case 3:
    if (ab == 0) {
      snprintf(dst,len+1,"%c%c%c",
	       b("1110-0000"),b("1010-0000"),b("1000-0000"));
    }
    if (ab == 1) {
      snprintf(dst,len+1,"%c%c%c",
	       b("1110-1111"),b("1011-1111"),b("1011-1111"));
    }
    break;  
  case 4:
    if (ab == 0) {
      snprintf(dst,len+1,"%c%c%c%c",
	       b("1111-0000"),b("1001-0000"),b("1000-0000"),b("1000-0000"));
    }
    if (ab == 1) {
      snprintf(dst,len+1,"%c%c%c%c",
	       b("1111-0111"),b("1011-1111"),b("1011-1111"),b("1011-1111"));
    }
    break;  
  }
}

int utf8testval(int len,int ab) {
  int val=-1;
  switch(len) {
  case 1:
    if (ab == 0) { val = 0x00; break; }
    if (ab == 1) { val = 0x7F; break; }
    break;
  case 2:
    if (ab == 0) { val = 0x80; break; }
    if (ab == 1) { val = 0x7FF; break; }
    break;  
  case 3:
    if (ab == 0) { val = 0x800; break; }
    if (ab == 1) { val = 0xFFFF; break; }
    break;  
  case 4:
    if (ab == 0) { val = 0x10000; break; }
    if (ab == 1) { val = 0x1FFFFF; break; }
    break;
  }
  return val;
}

FACTS(UTF8Len1) {
  unsigned xc,yc,len;
  char x8[1024],xtmp[1024],y8[1024],ytmp[1024];

  xc=0; yc=0x7F;
  snprintf(x8,sizeof(x8),"%c",b("0000-0000"));
  snprintf(y8,sizeof(y8),"%c",b("0111-1111"));

  len = 1;
  FACT(xc,==,UTF8_MINVAL[len-1]);
  if (len > 1) FACT(xc,==,UTF8_MAXVAL[len-2]+1);
  FACT(yc,==,UTF8_MAXVAL[len-1]);
  if (len < 4) FACT(yc,==,UTF8_MINVAL[len]-1);
  FACT(len,==,utf8declen(x8,len));
  FACT(-1,==,utf8declen(x8,len-1));
  FACT(xc,==,utf8decval(x8,len));
  FACT(len,==,utf8enclen(xc));
  utf8encval(xtmp,xc,len);
  FACT(strncmp(xtmp,x8,len),==,0);
  
  FACT(len,==,utf8declen(y8,len));
  FACT(-1,==,utf8declen(y8,len-1));  
  FACT(yc,==,utf8decval(y8,len));
  FACT(len,==,utf8enclen(yc));
  utf8encval(ytmp,yc,len);
  FACT(strncmp(ytmp,y8,len),==,0);
}

FACTS(UTF8Len2) {

  unsigned xc,yc,len;
  char x8[1024],xtmp[1024],y8[1024],ytmp[1024];

  xc=0x80; yc=0x7FF;
  snprintf(x8,sizeof(x8),"%c%c",b("1100-0010"),b("1000-0000"));
  snprintf(y8,sizeof(y8),"%c%c",b("1101-1111"),b("1011-1111"));

  len = 2;
  FACT(xc,==,UTF8_MINVAL[len-1]);
  if (len > 1) FACT(xc,==,UTF8_MAXVAL[len-2]+1);
  FACT(yc,==,UTF8_MAXVAL[len-1]);
  if (len < 4) FACT(yc,==,UTF8_MINVAL[len]-1);
  FACT(len,==,utf8declen(x8,len));
  FACT(-1,==,utf8declen(x8,len-1));
  FACT(xc,==,utf8decval(x8,len));
  FACT(len,==,utf8enclen(xc));
  utf8encval(xtmp,xc,len);
  FACT(strncmp(xtmp,x8,len),==,0);
  
  FACT(len,==,utf8declen(y8,len));
  FACT(-1,==,utf8declen(y8,len-1));  
  FACT(yc,==,utf8decval(y8,len));
  FACT(len,==,utf8enclen(yc));
  utf8encval(ytmp,yc,len);
  FACT(strncmp(ytmp,y8,len),==,0);
}

FACTS(UTF8Len3) {
  unsigned xc,yc,len;
  char x8[1024],xtmp[1024],y8[1024],ytmp[1024];

  xc=0x800; yc=0xFFFF;
  snprintf(x8,sizeof(x8),"%c%c%c",
	   b("1110-0000"),b("1010-0000"),b("1000-0000"));
  snprintf(y8,sizeof(y8),"%c%c%c",
	   b("1110-1111"),b("1011-1111"),b("1011-1111"));

  len = 3;
  FACT(xc,==,UTF8_MINVAL[len-1]);
  if (len > 1) FACT(xc,==,UTF8_MAXVAL[len-2]+1);
  FACT(yc,==,UTF8_MAXVAL[len-1]);
  if (len < 4) FACT(yc,==,UTF8_MINVAL[len]-1);
  FACT(len,==,utf8declen(x8,len));
  FACT(-1,==,utf8declen(x8,len-1));
  FACT(xc,==,utf8decval(x8,len));
  FACT(len,==,utf8enclen(xc));
  utf8encval(xtmp,xc,len);
  FACT(strncmp(xtmp,x8,len),==,0);
  
  FACT(len,==,utf8declen(y8,len));
  FACT(-1,==,utf8declen(y8,len-1));  
  FACT(yc,==,utf8decval(y8,len));
  FACT(len,==,utf8enclen(yc));
  utf8encval(ytmp,yc,len);
  FACT(strncmp(ytmp,y8,len),==,0);
}

FACTS(UTF8Len4) {
  unsigned xc,yc,len;
  char x8[1024],xtmp[1024],y8[1024],ytmp[1024];

  xc=0x10000; yc=0x1FFFFF;
  snprintf(x8,sizeof(x8),"%c%c%c%c",
	   b("1111-0000"),b("1001-0000"),b("1000-0000"),b("1000-0000"));
  snprintf(y8,sizeof(y8),"%c%c%c%c",
	   b("1111-0111"),b("1011-1111"),b("1011-1111"),b("1011-1111"));

  len = 4;
  FACT(xc,==,UTF8_MINVAL[len-1]);
  if (len > 1) FACT(xc,==,UTF8_MAXVAL[len-2]+1);
  FACT(yc,==,UTF8_MAXVAL[len-1]);
  if (len < 4) FACT(yc,==,UTF8_MINVAL[len]-1);
  FACT(len,==,utf8declen(x8,len));
  FACT(-1,==,utf8declen(x8,len-1));
  FACT(xc,==,utf8decval(x8,len));
  FACT(len,==,utf8enclen(xc));
  utf8encval(xtmp,xc,len);
  FACT(strncmp(xtmp,x8,len),==,0);
  
  FACT(len,==,utf8declen(y8,len));
  FACT(-1,==,utf8declen(y8,len-1));  
  FACT(yc,==,utf8decval(y8,len));
  FACT(len,==,utf8enclen(yc));
  utf8encval(ytmp,yc,len);
  FACT(strncmp(ytmp,y8,len),==,0);
}

FACTS(decode) {
  unsigned char u1[80],u2[80];
  uint32_t w1[80],w2[80];
  
  FACT(utf8decode("",0,NULL,0),==,0);
  for (int len=1; len<=4; ++len) {
    for (int ab=0; ab <= 1; ++ab) {
      for (int i=0; i<80; ++i) {
	u1[i]=u2[i]=-1;
	w1[i]=w2[i]=-1;
      }
      utf8teststr(u1,len,ab);
      int32_t wc32=utf8testval(len,ab);
      w1[0]=wc32;
      FACT(wc32,==,w1[0]);
      FACT(utf8decode(u1,len,NULL,0),==,1);
      FACT(utf8encode(w1,1,NULL,0),==,len);
      utf8decode(u1,len,w2,1);
      utf8encode(w1,1,u2,len);
      FACT(w1[0],==,w2[0]);
      for (int i=0; i<len; ++i) FACT(u1[i],==,u2[i]);
    }
  }
  for (int len1=1; len1<=4; ++len1) {
    for (int ab1=0; ab1 <= 1; ++ab1) {
      for (int len2=1; len2<=4; ++len2) {
	for (int ab2=0; ab2 <= 1; ++ab2) {
	  for (int i=0; i<80; ++i) {
	    u1[i]=u2[i]=-1;
	    w1[i]=w2[i]=-1;
	  }
	  utf8teststr(u1,len1,ab1);
	  utf8teststr(u1+len1,len2,ab2);
	  w1[0]=utf8testval(len1,ab1);
	  w1[1]=utf8testval(len2,ab2);
	  FACT(utf8decode(u1,len1+len2,NULL,0),==,2);
	  FACT(utf8encode(w1,2,NULL,0),==,len1+len2);
	  utf8decode(u1,len1+len2,w2,2);
	  utf8encode(w1,2,u2,len1+len2);
	  FACT(w1[0],==,w2[0]);
	  FACT(w1[1],==,w2[1]);	  
	  for (int i=0; i<len1+len2; ++i) FACT(u1[i],==,u2[i]);
	}
      }
    }
  }
}



FACTS(Suites) {
  uint32_t wc [] = {0x2663, 0x2666, 0x2665, 0x2660};  
  for (int i=0; i<4; ++i) {
    char buf[16];
    int utf8len = utf8enclen(wc[i]);
    utf8encval(buf,wc[i],utf8len);
    int utf8len2=utf8declen(buf,utf8len);
    FACT(utf8len2,==,utf8len);
    int wc2=utf8decval(buf,utf8len); 
    FACT(wc2,==,wc[i]);
    buf[utf8len]=0;
  }
}

FACTS_FINISHED
FACTS_MAIN
