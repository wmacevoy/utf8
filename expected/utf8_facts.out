src/utf8_facts.c 75: UTF8Len1 facts check started
src/utf8_facts.c 75: UTF8Len1 facts check ended
src/utf8_facts.c 103: UTF8Len2 facts check started
src/utf8_facts.c 103: UTF8Len2 facts check ended
src/utf8_facts.c 132: UTF8Len3 facts check started
src/utf8_facts.c 132: UTF8Len3 facts check ended
src/utf8_facts.c 162: UTF8Len4 facts check started
src/utf8_facts.c 162: UTF8Len4 facts check ended
src/utf8_facts.c 192: decode facts check started
src/utf8_facts.c 192: decode facts check ended
src/utf8_facts.c 242: Suits facts check started
src/utf8_facts.c 242: Suits facts check ended
facts summary.
facts check UTF8Len1 [1;32mpassed[0m
facts check UTF8Len2 [1;32mpassed[0m
facts check UTF8Len3 [1;32mpassed[0m
facts check UTF8Len4 [1;32mpassed[0m
facts check decode [1;32mpassed[0m
facts check Suits [1;32mpassed[0m
691 (100.0%) truths and 0 (0.0%) fictions checked.
